File: admin_platform/cpp/src/handlers/token_handler.cpp
/**
 * TigerWallet Admin Platform - C++ Token Handler Implementation
 * High-performance token management handler
 */

#include "../include/handlers/token_handler.h"
#include "../include/database.h"
#include "../include/redis_client.h"
#include <iostream>
#include <sstream>

namespace tiger {
namespace handlers {

TokenHandler::TokenHandler(std::shared_ptr<Database> db, std::shared_ptr<RedisClient> redis)
    : db_(db), redis_(redis) {}

nlohmann::json TokenHandler::get_tokens(const nlohmann::json& request) {
    int page = request.value("page", 1);
    int limit = request.value("limit", 20);
    std::string status = request.value("status", "");
    std::string chain = request.value("chain", "");
    std::string search = request.value("search", "");
    
    std::stringstream query;
    query << "SELECT id, token_id, name, symbol, contract_addr, decimals, "
          << "chain_id, chain_name, is_active, is_verified, is_native_token, "
          << "logo_url, website, price, price_change_24h, volume_24h, created_at "
          << "FROM tokens WHERE 1=1";
    
    if (!status.empty()) {
        query << " AND is_active = " << (status == "active" ? "true" : "false");
    }
    
    if (!chain.empty()) {
        query << " AND chain_id = '" << db_->escape_string(chain) << "'";
    }
    
    if (!search.empty()) {
        query << " AND (name ILIKE '%" << db_->escape_string(search) << "%' "
              << "OR symbol ILIKE '%" << db_->escape_string(search) << "%')";
    }
    
    query << " ORDER BY created_at DESC";
    query << " LIMIT " << limit << " OFFSET " << ((page - 1) * limit);
    
    auto result = db_->execute(query.str());
    
    if (!result) {
        return nlohmann::json::object({
            {"error", "Failed to fetch tokens"},
            {"success", false}
        });
    }
    
    nlohmann::json tokens = nlohmann::json::array();
    
    for (int i = 0; i < result->rows(); ++i) {
        tokens.push_back(nlohmann::json::object({
            {"id", result->get_value(i, "id")},
            {"token_id", result->get_value(i, "token_id")},
            {"name", result->get_value(i, "name")},
            {"symbol", result->get_value(i, "symbol")},
            {"contract_addr", result->get_value(i, "contract_addr")},
            {"decimals", std::stoi(result->get_value(i, "decimals"))},
            {"chain_id", result->get_value(i, "chain_id")},
            {"chain_name", result->get_value(i, "chain_name")},
            {"is_active", result->get_value(i, "is_active") == "true"},
            {"is_verified", result->get_value(i, "is_verified") == "true"},
            {"is_native_token", result->get_value(i, "is_native_token") == "true"},
            {"logo_url", result->get_value(i, "logo_url")},
            {"website", result->get_value(i, "website")},
            {"price", result->get_value(i, "price").empty() ? nullptr : std::stod(result->get_value(i, "price"))},
            {"price_change_24h", result->get_value(i, "price_change_24h").empty() ? nullptr : std::stod(result->get_value(i, "price_change_24h"))},
            {"volume_24h", result->get_value(i, "volume_24h").empty() ? nullptr : std::stod(result->get_value(i, "volume_24h"))},
            {"created_at", result->get_value(i, "created_at")}
        }));
    }
    
    // Get total count
    std::stringstream count_query;
    count_query << "SELECT COUNT(*) as total FROM tokens WHERE 1=1";
    auto count_result = db_->execute(count_query.str());
    int total = count_result && count_result->rows() > 0 ? 
        std::stoi(count_result->get_value(0, "total")) : 0;
    
    return nlohmann::json::object({
        {"success", true},
        {"data", tokens},
        {"meta", nlohmann::json::object({
            {"page", page},
            {"limit", limit},
            {"total", total},
            {"total_pages", (total + limit - 1) / limit}
        })}
    });
}

nlohmann::json TokenHandler::get_token(const std::string& token_id) {
    std::stringstream query;
    query << "SELECT id, token_id, name, symbol, contract_addr, decimals, "
          << "chain_id, chain_name, is_active, is_verified, is_native_token, "
          << "logo_url, website, price, price_change_24h, volume_24h, created_at "
          << "FROM tokens WHERE id = '" << db_->escape_string(token_id) << "'";
    
    auto result = db_->execute(query.str());
    
    if (!result || result->rows() == 0) {
        return nlohmann::json::object({
            {"error", "Token not found"},
            {"success", false}
        });
    }
    
    return nlohmann::json::object({
        {"success", true},
        {"data", nlohmann::json::object({
            {"id", result->get_value(0, "id")},
            {"token_id", result->get_value(0, "token_id")},
            {"name", result->get_value(0, "name")},
            {"symbol", result->get_value(0, "symbol")},
            {"contract_addr", result->get_value(0, "contract_addr")},
            {"decimals", std::stoi(result->get_value(0, "decimals"))},
            {"chain_id", result->get_value(0, "chain_id")},
            {"chain_name", result->get_value(0, "chain_name")},
            {"is_active", result->get_value(0, "is_active") == "true"},
            {"is_verified", result->get_value(0, "is_verified") == "true"},
            {"is_native_token", result->get_value(0, "is_native_token") == "true"},
            {"logo_url", result->get_value(0, "logo_url")},
            {"website", result->get_value(0, "website")},
            {"created_at", result->get_value(0, "created_at")}
        })}
    });
}

nlohmann::json TokenHandler::create_token(const nlohmann::json& data) {
    std::string token_id = "tk_" + std::to_string(std::time(nullptr)) + "_" + std::to_string(rand() % 10000);
    
    std::stringstream query;
    query << "INSERT INTO tokens (id, token_id, name, symbol, contract_addr, decimals, "
          << "chain_id, chain_name, is_active, is_verified, is_native_token, logo_url, website, created_at) "
          << "VALUES (gen_random_uuid(), '" << db_->escape_string(token_id) << "', "
          << "'" << db_->escape_string(data.value("name", "")) << "', "
          << "'" << db_->escape_string(data.value("symbol", "")) << "', "
          << "'" << db_->escape_string(data.value("contract_addr", "")) << "', "
          << data.value("decimals", 18) << ", "
          << "'" << db_->escape_string(data.value("chain_id", "")) << "', "
          << "'" << db_->escape_string(data.value("chain_name", "")) << "', "
          << "false, false, false, "
          << "'" << db_->escape_string(data.value("logo_url", "")) << "', "
          << "'" << db_->escape_string(data.value("website", "")) << "', "
          << "NOW())";
    
    auto result = db_->execute(query.str());
    
    if (!result) {
        return nlohmann::json::object({
            {"error", "Failed to create token"},
            {"success", false}
        });
    }
    
    return nlohmann::json::object({
        {"success", true},
        {"message", "Token created successfully"},
        {"data", nlohmann::json::object({
            {"token_id", token_id}
        })}
    });
}

nlohmann::json TokenHandler::update_token(const std::string& token_id, const nlohmann::json& data) {
    std::vector<std::string> updates;
    
    if (data.contains("name")) {
        updates.push_back("name = '" + db_->escape_string(data["name"].get<std::string>()) + "'");
    }
    if (data.contains("logo_url")) {
        updates.push_back("logo_url = '" + db_->escape_string(data["logo_url"].get<std::string>()) + "'");
    }
    if (data.contains("website")) {
        updates.push_back("website = '" + db_->escape_string(data["website"].get<std::string>()) + "'");
    }
    if (data.contains("is_active")) {
        updates.push_back("is_active = " + (data["is_active"].get<bool>() ? "true" : "false"));
    }
    
    if (updates.empty()) {
        return nlohmann::json::object({
            {"error", "No fields to update"},
            {"success", false}
        });
    }
    
    std::stringstream query;
    query << "UPDATE tokens SET " << updates[0];
    for (size_t i = 1; i < updates.size(); ++i) {
        query << ", " << updates[i];
    }
    query << ", updated_at = NOW() WHERE id = '" << db_->escape_string(token_id) << "'";
    
    auto result = db_->execute(query.str());
    
    if (!result) {
        return nlohmann::json::object({
            {"error", "Failed to update token"},
            {"success", false}
        });
    }
    
    // Invalidate cache
    if (redis_) {
        redis_->del("token:" + token_id);
    }
    
    return nlohmann::json::object({
        {"success", true},
        {"message", "Token updated successfully"}
    });
}

nlohmann::json TokenHandler::delete_token(const std::string& token_id) {
    std::stringstream query;
    query << "DELETE FROM tokens WHERE id = '" << db_->escape_string(token_id) << "'";
    
    auto result = db_->execute(query.str());
    
    if (!result) {
        return nlohmann::json::object({
            {"error", "Failed to delete token"},
            {"success", false}
        });
    }
    
    // Invalidate cache
    if (redis_) {
        redis_->del("token:" + token_id);
    }
    
    return nlohmann::json::object({
        {"success", true},
        {"message", "Token deleted successfully"}
    });
}

nlohmann::json TokenHandler::verify_token(const std::string& token_id) {
    std::stringstream query;
    query << "UPDATE tokens SET is_verified = true, verified_at = NOW() "
          << "WHERE id = '" << db_->escape_string(token_id) << "'";
    
    auto result = db_->execute(query.str());
    
    if (!result) {
        return nlohmann::json::object({
            {"error", "Failed to verify token"},
            {"success", false}
        });
    }
    
    // Invalidate cache
    if (redis_) {
        redis_->del("token:" + token_id);
    }
    
    return nlohmann::json::object({
        {"success", true},
        {"message", "Token verified successfully"}
    });
}

} // namespace handlers
} // namespace tiger

File: admin/go/internal/handlers/token_handler.go
package handlers

import (
	"net/http"
	"strconv"
	"time"

	"github.com/tigerwallet/admin/internal/models"
	"github.com/tigerwallet/admin/pkg/auth"
	"github.com/tigerwallet/admin/pkg/database"

	"github.com/gin-gonic/gin"
)

// TokenHandler handles token-related requests
type TokenHandler struct {
	db      *database.PostgresDB
	authSvc *auth.AuthService
}

// NewTokenHandler creates a new token handler
func NewTokenHandler(db *database.PostgresDB, authSvc *auth.AuthService) *TokenHandler {
	return &TokenHandler{
		db:      db,
		authSvc: authSvc,
	}
}

// ListTokens lists all tokens
func (h *TokenHandler) ListTokens(c *gin.Context) {
	page := c.DefaultQuery("page", "1")
	pageSize := c.DefaultQuery("page_size", "20")
	search := c.Query("search")
	chain := c.Query("chain")
	status := c.Query("status")

	pageInt, _ := strconv.Atoi(page)
	pageSizeInt, _ := strconv.Atoi(pageSize)

	var tokens []models.Token
	var total int64

	query := h.db.Model(&models.Token{})

	if search != "" {
		query = query.Where("name ILIKE ? OR symbol ILIKE ?", "%"+search+"%", "%"+search+"%")
	}
	if chain != "" {
		query = query.Where("chain = ?", chain)
	}
	if status != "" {
		if status == "active" {
			query = query.Where("is_active = ?", true)
		} else if status == "inactive" {
			query = query.Where("is_active = ?", false)
		}
	}

	query.Count(&total)

	offset := (pageInt - 1) * pageSizeInt
	query = query.Offset(offset).Limit(pageSizeInt).Order("created_at DESC")

	if err := query.Find(&tokens).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch tokens"})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"data":        tokens,
		"total":       total,
		"page":        pageInt,
		"page_size":   pageSizeInt,
		"total_pages": (total + int64(pageSizeInt) - 1) / int64(pageSizeInt),
	})
}

// GetToken gets a token by ID
func (h *TokenHandler) GetToken(c *gin.Context) {
	tokenID := c.Param("id")

	var token models.Token
	if err := h.db.First(&token, tokenID).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Token not found"})
		return
	}

	c.JSON(http.StatusOK, token)
}

// CreateToken creates a new token
func (h *TokenHandler) CreateToken(c *gin.Context) {
	adminID := c.GetUint("admin_id")

	var req struct {
		Name            string `json:"name" binding:"required"`
		Symbol          string `json:"symbol" binding:"required"`
		ContractAddress string `json:"contract_address"`
		Chain           string `json:"chain" binding:"required"`
		Decimals        int    `json:"decimals"`
		TotalSupply     string `json:"total_supply"`
		LogoURL         string `json:"logo_url"`
		Website         string `json:"website"`
		Description     string `json:"description"`
		ListingFee      string `json:"listing_fee"`
	}

	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid request body"})
		return
	}

	// Check if token symbol already exists
	var existingToken models.Token
	if err := h.db.Where("symbol = ? AND chain = ?", req.Symbol, req.Chain).First(&existingToken).Error; err == nil {
		c.JSON(http.StatusConflict, gin.H{"error": "Token with this symbol already exists on this chain"})
		return
	}

	if req.Decimals == 0 {
		req.Decimals = 18
	}

	now := time.Now()
	token := models.Token{
		Name:            req.Name,
		Symbol:          req.Symbol,
		ContractAddress: req.ContractAddress,
		Chain:           req.Chain,
		Decimals:        req.Decimals,
		TotalSupply:     req.TotalSupply,
		LogoURL:         req.LogoURL,
		Website:         req.Website,
		Description:     req.Description,
		IsActive:        true,
		IsVerified:      false,
		ListingFee:      req.ListingFee,
		ListedBy:        adminID,
		ListedAt:        &now,
	}

	if err := h.db.Create(&token).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to create token"})
		return
	}

	// Log activity
	logAdminActivity(h.db, adminID, "create_token", "token", strconv.FormatUint(uint64(token.ID), 10), "Created token: "+token.Symbol, c.ClientIP(), c.Request.UserAgent())

	c.JSON(http.StatusCreated, token)
}

// UpdateToken updates a token
func (h *TokenHandler) UpdateToken(c *gin.Context) {
	adminID := c.GetUint("admin_id")
	tokenID := c.Param("id")

	var req struct {
		Name        string `json:"name"`
		LogoURL     string `json:"logo_url"`
		Website     string `json:"website"`
		Description string `json:"description"`
		Price       string `json:"price"`
		IsActive    *bool  `json:"is_active"`
		IsVerified  *bool  `json:"is_verified"`
	}

	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid request body"})
		return
	}

	var token models.Token
	if err := h.db.First(&token, tokenID).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Token not found"})
		return
	}

	updates := map[string]interface{}{}

	if req.Name != "" {
		updates["name"] = req.Name
	}
	if req.LogoURL != "" {
		updates["logo_url"] = req.LogoURL
	}
	if req.Website != "" {
		updates["website"] = req.Website
	}
	if req.Description != "" {
		updates["description"] = req.Description
	}
	if req.Price != "" {
		updates["price"] = req.Price
	}
	if req.IsActive != nil {
		updates["is_active"] = *req.IsActive
	}
	if req.IsVerified != nil {
		updates["is_verified"] = *req.IsVerified
	}

	if err := h.db.Model(&token).Updates(updates).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to update token"})
		return
	}

	// Log activity
	logAdminActivity(h.db, adminID, "update_token", "token", tokenID, "Updated token: "+token.Symbol, c.ClientIP(), c.Request.UserAgent())

	c.JSON(http.StatusOK, token)
}

// DeleteToken soft deletes a token
func (h *TokenHandler) DeleteToken(c *gin.Context) {
	adminID := c.GetUint("admin_id")
	tokenID := c.Param("id")

	var token models.Token
	if err := h.db.First(&token, tokenID).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Token not found"})
		return
	}

	// Soft delete - deactivate
	if err := h.db.Model(&token).Update("is_active", false).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to delete token"})
		return
	}

	// Log activity
	logAdminActivity(h.db, adminID, "delete_token", "token", tokenID, "Deleted token: "+token.Symbol, c.ClientIP(), c.Request.UserAgent())

	c.JSON(http.StatusOK, gin.H{"message": "Token deleted successfully"})
}

// ActivateToken activates a token
func (h *TokenHandler) ActivateToken(c *gin.Context) {
	adminID := c.GetUint("admin_id")
	tokenID := c.Param("id")

	var token models.Token
	if err := h.db.First(&token, tokenID).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Token not found"})
		return
	}

	if err := h.db.Model(&token).Update("is_active", true).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to activate token"})
		return
	}

	// Log activity
	logAdminActivity(h.db, adminID, "activate_token", "token", tokenID, "Activated token: "+token.Symbol, c.ClientIP(), c.Request.UserAgent())

	c.JSON(http.StatusOK, gin.H{"message": "Token activated successfully"})
}

// DeactivateToken deactivates a token
func (h *TokenHandler) DeactivateToken(c *gin.Context) {
	adminID := c.GetUint("admin_id")
	tokenID := c.Param("id")

	var token models.Token
	if err := h.db.First(&token, tokenID).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Token not found"})
		return
	}

	if err := h.db.Model(&token).Update("is_active", false).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to deactivate token"})
		return
	}

	// Log activity
	logAdminActivity(h.db, adminID, "deactivate_token", "token", tokenID, "Deactivated token: "+token.Symbol, c.ClientIP(), c.Request.UserAgent())

	c.JSON(http.StatusOK, gin.H{"message": "Token deactivated successfully"})
}

// VerifyToken verifies a token
func (h *TokenHandler) VerifyToken(c *gin.Context) {
	adminID := c.GetUint("admin_id")
	tokenID := c.Param("id")

	var token models.Token
	if err := h.db.First(&token, tokenID).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Token not found"})
		return
	}

	if err := h.db.Model(&token).Update("is_verified", true).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to verify token"})
		return
	}

	// Log activity
	logAdminActivity(h.db, adminID, "verify_token", "token", tokenID, "Verified token: "+token.Symbol, c.ClientIP(), c.Request.UserAgent())

	c.JSON(http.StatusOK, gin.H{"message": "Token verified successfully"})
}

// GetTokenStats gets token statistics
func (h *TokenHandler) GetTokenStats(c *gin.Context) {
	var stats struct {
		TotalTokens      int64 `json:"total_tokens"`
		ActiveTokens     int64 `json:"active_tokens"`
		InactiveTokens   int64 `json:"inactive_tokens"`
		VerifiedTokens   int64 `json:"verified_tokens"`
		UnverifiedTokens int64 `json:"unverified_tokens"`
	}

	h.db.Model(&models.Token{}).Count(&stats.TotalTokens)
	h.db.Model(&models.Token{}).Where("is_active = ?", true).Count(&stats.ActiveTokens)
	h.db.Model(&models.Token{}).Where("is_active = ?", false).Count(&stats.InactiveTokens)
	h.db.Model(&models.Token{}).Where("is_verified = ?", true).Count(&stats.VerifiedTokens)
	h.db.Model(&models.Token{}).Where("is_verified = ?", false).Count(&stats.UnverifiedTokens)

	c.JSON(http.StatusOK, stats)
}

// UpdateTokenPrice updates token price
func (h *TokenHandler) UpdateTokenPrice(c *gin.Context) {
	adminID := c.GetUint("admin_id")
	tokenID := c.Param("id")

	var req struct {
		Price          string `json:"price" binding:"required"`
		MarketCap      string `json:"market_cap"`
		Volume24h      string `json:"volume_24h"`
		PriceChange24h string `json:"price_change_24h"`
	}

	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Price is required"})
		return
	}

	var token models.Token
	if err := h.db.First(&token, tokenID).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Token not found"})
		return
	}

	updates := map[string]interface{}{
		"price": req.Price,
	}

	if req.MarketCap != "" {
		updates["market_cap"] = req.MarketCap
	}
	if req.Volume24h != "" {
		updates["volume_24h"] = req.Volume24h
	}
	if req.PriceChange24h != "" {
		updates["price_change_24h"] = req.PriceChange24h
	}

	if err := h.db.Model(&token).Updates(updates).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to update token price"})
		return
	}

	// Log activity
	logAdminActivity(h.db, adminID, "update_token_price", "token", tokenID, "Updated price: "+req.Price, c.ClientIP(), c.Request.UserAgent())

	c.JSON(http.StatusOK, token)
}
